README
NAME
    CHI::Config - Define CHI configuration outside your code

VERSION
    version 0.001002

SYNOPSIS
      use CHI::Config;

      my $config = CHI::Config->new(
        defaults => [
          # Defaults indeed has to mimic the source file for future purposes
          # ie: I plan to make some objects ( such as serializers )
          # be configurable as well and they can't really be defined as-is in JSON
          {
              type => 'driver',
              name => 'myproject.roflmayo',
              config => {
                  # Arguments to CHI->new()
              },
          },
        ],
      );

      my $cache = $config->get_cache('myproject.roflmayo');

      # Do stuff with $cache and get default behaviour

      # User creates ~/.chi/config.json
      [
        {
          'type' : 'driver',
          'name' : 'myproject.roflmayo',
          'config': {
             # CHI CONFIG HERE
          },
        }
      ]

      my $cache = $config->get_cache('myproject.roflmayo');  # Now gets user defined copy

CONSTRUCTOR ARGUMENTS
  "config_paths"
    *Optional*: An ArrayRef of path prefixes to scan and load.

    For instance:

      ( config_paths => ['./foo'] )

    Would automatically attempt to load any files called

      foo.yml
      foo.json
      foo.ini

    And load them with the relevant helpers.

    See "Config::Any" for details on this mechanism.

    Paths will be interpreted in the order specified, with the first one
    taking precedence over the latter ones for any given driver name, with
    "defaults" being taken only if they're needed.

    Default paths loaded are as follows:

        $ENV{CHI_CONFIG_DIR}/config.*
        ./chi_config.*
        ~/.chi/config.*
        /etc/chi/config.*

  "config_files"
    *Optional*: An ArrayRef of files to scan and load.

    If specified, this list entirely overrules that provided by
    "config_paths"

  "defaults"
    *Recommended*: An ArrayRef of defaults in the same notation as the
    configuration spec.

      defaults => [
           $entry,
           $entry,
           $entry,
      ],

    See "ENTRIES"

METHODS
  "get_cache"
    Retrieve an instance of a cache object for consumption.

      my $cache = $config->get_cache('myproject.myname');

      $cache-># things with CHI

ENTRIES
    Both the internal array based interface and the configuration file are a
    list of "Entries". Design somewhat inspired by "Config::MVP"'s sequence
    model, but much more lightweight.

  "driver" entry
    These make up the core of a configuration.

      {
        type => 'driver',

        # The following are all passed through to
        # CHI::Config::Driver

        # STRONGLY recommended
        name => 'mynamespace.mycachename',

        # RAW CHI arguments
        config => {
          %CONFIG    #
        },

        # return singleton or new caches?
        memoize => 0,
      }

    See "CHI::Config::Driver" for details.

  "version" entry
    This is a mostly unnecessary element simply designed to give some kind
    of informal API in the event there are changes in how the configuration
    is parsed.

    Currently, Spec version is == 0.1.0

      {
        type => 'version',

        # Declare a minimum version of CHI::Config
        min => 0.001000,

        # Declare a maximum version of CHI::Config
        max => 1.000000,

        # Require exactly specification 0.1.0
        spec => '0.1.0',
      }

    "max" and "min" give range controls on the version of "CHI::Config"
    itself.

    "spec" gives an exact match on the *interface* provided by
    "CHI::Config", and is processed as an exact string match.

    Any of the criteria not being satisfied will result in a "croak"

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for CHI-Config

0.001002 2014-11-08T02:39:29Z 57e6bad

0.001001 2014-11-08T02:38:47Z de26588
 [Bugfix]
 - Don't have a broken default on old Moo

0.001000 2014-11-06T22:23:22Z a824103
 - First version.



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.024.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Define CHI configuration outside your code",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "CHI-Config",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "CHI::Config",
  "PREREQ_PM" => {
    "CHI" => 0,
    "Carp" => 0,
    "Config::Any" => 0,
    "File::HomeDir" => 0,
    "Moo" => 0,
    "Storable" => 0,
    "constant" => 0,
    "strict" => 0,
    "version" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Path::Tiny" => 0,
    "Test::More" => "0.89",
    "Test::TempDir::Tiny" => 0
  },
  "VERSION" => "0.001002",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "CHI" => 0,
  "Carp" => 0,
  "Config::Any" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::HomeDir" => 0,
  "File::Spec" => 0,
  "Moo" => 0,
  "Path::Tiny" => 0,
  "Storable" => 0,
  "Test::More" => "0.89",
  "Test::TempDir::Tiny" => 0,
  "constant" => 0,
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/CHI/Config/Driver.pm
use 5.006;
use strict;
use warnings;

package CHI::Config::Driver;

our $VERSION = '0.001002'; # TRIAL

# ABSTRACT: Container for Driver configuration

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo qw( has );

has 'name'     => ( is => 'ro', required => 1 );
has 'file'     => ( is => 'ro', required => 1 );
has 'entry_no' => ( is => 'ro', required => 1 );
has 'config'   => ( is => 'ro', required => 1 );
has 'memoize'  => ( is => 'ro', lazy     => 1, default => sub { undef } );

sub get_cache {
  my ($self) = @_;
  return $self->{_cache} if exists $self->{_cache};

  require CHI;
  require Storable;

  my $instance = CHI->new( %{ Storable::dclone( $self->config ) } );

  $self->{_cache} = $instance if $self->memoize;

  return $instance;
}

sub source {
  my ($self) = @_;
  return sprintf '%s ( entry #%s )', $self->file, $self->entry_no;
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

CHI::Config::Driver - Container for Driver configuration

=head1 VERSION

version 0.001002

=head1 SYNOPSIS

  my $config = CHI::Config::Driver->new(
    name     => 'my.name.here',
    file     => '/path/config_was_loaded_from',    # for debugging
    entry_no => 5,                                 # entry number in 5 for debugging
    config   => {},                                # CHI Payload
    memoize  => 0 || 1,                            # whether multiple calls make singular objects
  );

This is mostly an implementation detail for C<CHI::Config>.

Each instance directly maps to an entry in the configuration file.

Two parameters C<file> and C<entry_no> are provided automatically by the infrastructure.

=head1 METHODS

=head2 C<get_cache>

Returns a C<CHI> cache object using the definition in C<config>.

This may return the same object across multiple calls ( if C< L</memoize> > is true ).

Otherwise returns a new C<CHI> instance with every call.

=head2 C<source>

Returns a string

=head1 ATTRIBUTES

=head2 C<name>

The name of this driver for look-up purposes.

B<Strongly> Recommended format ( I<especially> for CPAN ) is some sort
of alias path with a prefix, i.e:

  name => 'dist-zilla-cache.www-objects'

=head2 C<config>

This is a raw hash payload that gets passed to C<< CHI->new() >>.

  config => {
    driver => 'Memory',
    ...                   # For example
  }

=head2 C<memoize>

This attribute dictates whether multiple calls to C<get_cache> will return
I<different> cache objects ( which may happen to share implementation details ),
or whether it will return a I<single> cache object for multiple calls.

  memoize => 0  # Default, each call to get_cache returns a new CHI instance
  memoize => 1  # Construct CHI cache once and return it multiple times.

=head2 C<file>

Pertains to the location this C<Driver> configuration was determined from.

This is an implementation detail used for debug purposes in C<CHI::Config>
and is normally provided by the various configuration sourcing tools.

=head2 C<entry_no>

Pertains to the placement in the array of configuration details ( in C<file> )
where the configuration for this C<Driver> was found.

This is an implementation detail used for debug purposes in C<CHI::Config>
and is normally provided by the various configuration sourcing tools.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw(CPAN ArrayRef API));
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
$bundle->remove_policy('CodeLayout::RequireUseUTF8');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



lib/CHI/Config/DriverPool.pm
use 5.006;
use strict;
use warnings;

package CHI::Config::DriverPool;

our $VERSION = '0.001002'; # TRIAL

# ABSTRACT: A Collection of Driver definitions

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo qw( has );
use Carp qw( carp croak );

## no critic (ValuesAndExpressions::ProhibitConstantPragma)
use constant DEBUGGING => $ENV{CHI_CONFIG_DEBUG};

has '_drivers' => ( is => ro =>, lazy => 1, default => sub { {} } );

sub add_driver {
  my ( $self, %config ) = @_;

  require CHI::Config::Driver;

  my $instance = CHI::Config::Driver->new(%config);
  my $name     = $instance->name;

  if ( not exists $self->_drivers->{$name} ) {
    $self->_drivers->{$name} = $instance;
    return;
  }
  return unless DEBUGGING;
  ## Shadowing should be default anyway.
  my $template = '%s: %s';
  my $old      = sprintf $template, '   Kept', $self->_drivers->{$name}->source;
  my $new      = sprintf $template, 'Ignored', $instance->source;
  carp "Duplicate Driver definition ignored\n\t$old\n\t$new\n";
  return;
}

sub get_driver {
  my ( $self, $name ) = @_;
  return $self->_drivers->{$name} if exists $self->_drivers->{$name};
  croak "No default for driver <$name> defined and none specified in configuration";
}

sub get_cache {
  my ( $self, $name ) = @_;
  return $self->get_driver($name)->get_cache;
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

CHI::Config::DriverPool - A Collection of Driver definitions

=head1 VERSION

version 0.001002

=head1 SYNOPSIS

  use CHI::Config::DriverPool;

  my $dp = CHI::Config::DriverPool->new();

  $dp->add_driver(
    name => ...    # See CHI::Config::Driver for arguments
  );

  $dp->get_driver($name);    # Fetches a CHI::Config::Driver instance

  $dp->get_cache($name);     # Shorthand for $dp->get_driver($name)->get_cache

=head1 DESCRIPTION

This container serves to aggregate configuration of all driver
configurations across multiple files.

It is assumed that the consumer is traversing a tree of configurations of
some kind in order of "most relevant" to "least relevant", for example,
"most relevant" would be the files in C<./> , second-most relevant
would be C<~/> somewhere, somewhere later is C</etc/> and least relevant
is any defaults provided by the program itself.

As such, the first definition seen of any given name is considered
"the one" to be used and subsequent ones are ignored.

=head1 METHODS

=head2 C<add_driver>

Create and inject a unique driver configuration in the dictionary.

First come first served.

  $pool->add_driver( %config );

Arguments for C<%config> are the same as for L<< C<< CHI::Config::Driver->new() >>|CHI::Config::Driver >>.

=head2 C<get_driver>

Fetch a named driver configuration from the dictionary.

  $pool->get_driver( $name );

This will return a L<< C<CHI::Config::Driver>|CHI::Config::Driver >> instance, or C<croak> if one does not exist.

=head2 C<get_cache>

Fetch and a named C<CHI> cache object by requesting one from the named C<CHI::Config::Driver> object.

=head1 ENVIRONMENT

=head2 CHI_CONFIG_DEBUG

Enables reporting where driver definitions are being shadowed.

This will show shadowing everywhere there is user defined configuration
in addition to a default, so is potentially more noisy than intended,
but it is hopefully useful in diagnosing misplaced identifiers.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/CHI/Config.pm
use 5.006;
use strict;
use warnings;

package CHI::Config;

our $VERSION = '0.001002'; # TRIAL

# ABSTRACT: Define CHI configuration outside your code

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Carp qw( croak );
use Moo qw( has around );

has '_constructor_caller' => (
  is       => 'ro',
  required => 1,
);

has '_config_paths' => (
  init_arg => 'config_paths',
  is       => 'ro',
  lazy     => 1,
  builder  => '_build_config_paths',
);

has '_config_files' => (
  init_arg => 'config_files',
  is       => 'ro',
  lazy     => 1,
  builder  => '_build_config_files',
);

has '_config' => (
  init_arg => undef,
  is       => 'ro',
  lazy     => 1,
  builder  => '_build_config',
);

has '_defaults' => (
  init_arg => 'defaults',
  is       => ro =>,
  lazy     => 1,
  builder  => '_build_defaults',
);

has '_drivers' => (
  is       => 'ro',
  init_arg => undef,
  lazy     => 1,
  builder  => '_build_drivers',
  handles  => {
    '_add_driver' => 'add_driver',
    '_get_driver' => 'get_driver',
    'get_cache'   => 'get_cache',
  },
);

sub BUILDARGS {
  my ( undef, @args ) = @_;
  my (@caller) = caller 1;    # +1 for Moo
  if ( 1 == @args and ref $args[0] ) {
    $args[0]->{_constructor_caller} = $caller[1];
    return $args[0];
  }
  return { @args, _constructor_caller => $caller[1] };
}

sub BUILD {
  my ($self) = @_;
  ## no critic (Variables::ProhibitPackageVars)
  local $Carp::CarpLevel = $Carp::CarpLevel + 1;
  $self->_load_config;
  $self->_load_defaults;
  return;
}

sub _build_config_paths {
  require File::HomeDir;
  my (@scan_paths) = (
    ( exists $ENV{CHI_CONFIG_DIR} ? $ENV{CHI_CONFIG_DIR} . '/config' : () ),    # From ENV
    './chi_config',                                                             # ./
    File::HomeDir->my_home . '/.chi/config',                                    # ~/.chi/
    '/etc/chi/config',                                                          # /etc/chi/
  );
  return \@scan_paths;
}

sub _build_config_files { return [] }

sub _build_config {
  my ($self) = @_;
  require Config::Any;
  my %extras = ( use_ext => 1 );
  return Config::Any->load_files( { files => $self->_config_files, %extras } ) if @{ $self->_config_files };
  return Config::Any->load_stems( { stems => $self->_config_paths, %extras } );
}

sub _build_defaults {
  return [];
}

sub _build_drivers {
  require CHI::Config::DriverPool;
  return CHI::Config::DriverPool->new();
}

sub _load_version {
  my ( undef, %entry ) = @_;

  my $spec      = '0.1.0';
  my $entry_msg = "$entry{file} ( entry #$entry{entry_no} )";

  if ( $entry{spec} and $entry{spec} ne $spec ) {
    croak "Spec version required by $entry_msg is ==$entry{spec}, this is $spec";
  }
  if ( $entry{min} ) {
    require version;
    my $min = version->parse( $entry{min} );
    if ( $min > $VERSION ) {
      croak "Minimum version required by $entry_msg is $min, we have $VERSION";
    }
  }
  if ( $entry{max} ) {
    require version;
    my $max = version->parse( $entry{max} );
    if ( $max < $VERSION ) {
      croak "Maximum version required by $entry_msg is $max, we have $VERSION";
    }
  }
  return;
}

sub _load_entry {
  my ( $self, %entry ) = @_;
  my $context = sprintf q[%s ( entry #%s )], $entry{file}, $entry{entry_no};
  my $name = ( $entry{name} ? q[ named ] . $entry{name} : q[ ] );
  croak "No type specified for entry$name in $context" unless defined $entry{type};
  return $self->_load_version(%entry) if 'version' eq $entry{type};
  return $self->_add_driver(%entry)   if 'driver' eq $entry{type};
  croak "Unknown type $entry{type} in $context";
}

sub _load_array {
  my ( $self, $array, $file ) = @_;
  if ( not 'ARRAY' eq ref $array ) {
    return croak "Payload in $file should be an ARRAY of HASH";
  }
  my $entry_no = 1;
  for my $entry ( @{$array} ) {
    if ( not 'HASH' eq ref $entry ) {
      return croak "Entry $entry_no in $file is not a HASH";
    }
    $self->_load_entry(
      %{$entry},
      file     => $file,        #
      entry_no => $entry_no,    #
    );
    $entry_no++;
  }
  return;
}

sub _load_config {
  my ( $self, ) = @_;

  # Load from config first.
  for my $result ( @{ $self->_config } ) {
    for my $file ( keys %{$result} ) {
      $self->_load_array( $result->{$file}, $file );
    }
  }
  return;
}

sub _load_defaults {
  my ( $self, ) = @_;
  return $self->_load_array( $self->_defaults, $self->_constructor_caller );
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

CHI::Config - Define CHI configuration outside your code

=head1 VERSION

version 0.001002

=head1 SYNOPSIS

  use CHI::Config;

  my $config = CHI::Config->new(
    defaults => [
      # Defaults indeed has to mimic the source file for future purposes
      # ie: I plan to make some objects ( such as serializers )
      # be configurable as well and they can't really be defined as-is in JSON
      {
          type => 'driver',
          name => 'myproject.roflmayo',
          config => {
              # Arguments to CHI->new()
          },
      },
    ],
  );

  my $cache = $config->get_cache('myproject.roflmayo');

  # Do stuff with $cache and get default behaviour

  # User creates ~/.chi/config.json
  [
    {
      'type' : 'driver',
      'name' : 'myproject.roflmayo',
      'config': {
         # CHI CONFIG HERE
      },
    }
  ]

  my $cache = $config->get_cache('myproject.roflmayo');  # Now gets user defined copy

=head1 CONSTRUCTOR ARGUMENTS

=head2 C<config_paths>

I<Optional>: An ArrayRef of path prefixes to scan and load.

For instance:

  ( config_paths => ['./foo'] )

Would automatically attempt to load any files called

  foo.yml
  foo.json
  foo.ini

And load them with the relevant helpers.

See L<< C<Config::Any>|Config::Any >> for details on this mechanism.

Paths will be interpreted in the order specified, with the first one
taking precedence over the latter ones for any given driver name,
with C<defaults> being taken only if they're needed.

Default paths loaded are as follows:

    $ENV{CHI_CONFIG_DIR}/config.*
    ./chi_config.*
    ~/.chi/config.*
    /etc/chi/config.*

=head2 C<config_files>

I<Optional>: An ArrayRef of files to scan and load.

If specified, this list entirely overrules that provided by
L<< C<config_paths>|/config_paths >>

=head2 C<defaults>

I<Recommended>: An ArrayRef of defaults in the same notation as the configuration spec.

  defaults => [
       $entry,
       $entry,
       $entry,
  ],

See L</ENTRIES>

=head1 METHODS

=head2 C<get_cache>

Retrieve an instance of a cache object for consumption.

  my $cache = $config->get_cache('myproject.myname');

  $cache-># things with CHI

=for Pod::Coverage BUILD BUILDARGS

=head1 ENTRIES

Both the internal array based interface and the configuration file
are a list of C<Entries>. Design somewhat inspired by C<Config::MVP>'s
sequence model, but much more lightweight.

=head2 C<driver> entry

These make up the core of a configuration.

  {
    type => 'driver',

    # The following are all passed through to
    # CHI::Config::Driver

    # STRONGLY recommended
    name => 'mynamespace.mycachename',

    # RAW CHI arguments
    config => {
      %CONFIG    #
    },

    # return singleton or new caches?
    memoize => 0,
  }

See L<< C<CHI::Config::Driver>|CHI::Config::Driver >> for details.

=head2 C<version> entry

This is a mostly unnecessary element simply designed to give
some kind of informal API in the event there are changes in
how the configuration is parsed.

Currently, Spec version is == C<0.1.0>

  {
    type => 'version',

    # Declare a minimum version of CHI::Config
    min => 0.001000,

    # Declare a maximum version of CHI::Config
    max => 1.000000,

    # Require exactly specification 0.1.0
    spec => '0.1.0',
  }

C<max> and C<min> give range controls on the version of C<CHI::Config> itself.

C<spec> gives an exact match on the I<interface> provided by C<CHI::Config>, and is processed as an exact string match.

Any of the criteria not being satisfied will result in a C<croak>

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
